lucene/core/src/java/org/apache/lucene/search/FilteringLeafFieldComparator.java

```diff
@@ -32,8 +33,22 @@
   DocIdSetIterator competitiveIterator() throws IOException;
 
   /**
-   * Informs this leaf comparator that it is allowed to start updating its competitive iterator.
-   * This method is called from a collector when queue becomes full and threshold is reached.
+   * Informs this leaf comparator that hits threshold is reached.
+   * This method is called from a collector when hits threshold is reached.
+   * For some filtering comparators (e.g. {@code FilteringDocLeafComparator} reaching
+   * hits threshold is enough to start updating their iterators, even when queue is not yet full.
    */
-  void setCanUpdateIterator() throws IOException;
+  void setHitsThresholdReached() throws IOException;
+
+  /**
+   * Informs this leaf comparator that queue has become full.
+   * This method is called from a collector when queue becomes full.
+   */
+  void setQueueFull() throws IOException;
```

Comment:
That seems redundant. I think we can add a single method: setHitsThresholdReached(int). We have all the other contexts needed in the LeafFieldComparator. The complexity is that `setBottom `and `setHitsThresholdReached(int)` are called on the leaves even though they are global events. Maybe forcing the hierarchy of the filtering fields to use a `SimpleFieldComparator` could simplify things ? This way you don't have to repeat redundant informations (hits threshold and queue full) on each leave.

Reply:
addressed in 746c8fa